src/sync.rs

```diff
@@ -516,14 +516,36 @@ pub trait OnceLockExt<T>: once_lock_ext_sealed::Sealed {
         F: FnOnce() -> T;
 }
 
-struct Guard(*mut crate::ffi::PyThreadState);
+/// Extension trat for [`std::sync::Mutex`] which helps avoid deadlocks between
+/// the Python interpreter and acquiring the `Mutex`.
+pub trait MutexExt<T>: Sealed {
+    /// Lock this `Mutex` in a manner that cannot deadlock with the Python interpreter.
+    ///
+    /// Before attempting to lock the mutex, this function detaches from the
+    /// Python runtime. When the lock is acquired, it re-attaches to the Python
+    /// runtime before returning the `LockResult`. This avoids deadlocks between
+    /// the GIL and other global synchronization events triggered by the Python
+    /// interpreter.
+    fn lock_py_attached(
+        &self,
+        py: Python<'_>,
+    ) -> std::sync::LockResult<std::sync::MutexGuard<'_, T>>;
+}
+
+struct ThreadStateGuard(*mut crate::ffi::PyThreadState);
 
-impl Drop for Guard {
+impl Drop for ThreadStateGuard {
     fn drop(&mut self) {
         unsafe { ffi::PyEval_RestoreThread(self.0) };
     }
 }
 
+impl ThreadStateGuard {
+    fn new() -> ThreadStateGuard {
```

Comment:
Should we take the `py` token here, to proof that we are currently attached to the runtime (which is the safety invariant here, right)?

Reply:
Good point. I also added some docs with additional safety cautions for using `ThreadStateGuard`.

Reply:
Also, isn't this basically the implementation of `Python::allow_threads`? Can't we just use that?

Reply:
`MutexGuard` isn't `Send`, so I can't return it from `py.allow_threads`. The other uses in `OnceExt` and `OnceLockExt` can't use `allow_threads` either because they need to explicitly drop the guard before possibly calling into the C API.

I could maybe make it so that `ThreadStateGuard::new()` is `unsafe` to force users to think about where and when the calling thread is attached to the runtime. I don't think there's a way to safely prevent someone from using the `py` token while the thread is detached from the runtime.

Reply:
Ah right, I guess it would work for `MutexGuard` if `Ungil` would not rely on `Send` (like on nightly), but you're right, for `Once` it does work. 

> I don't think there's a way to safely prevent someone from using the `py` token while the thread is detached from the runtime.

Yeah, I don't see one either. Given this an internal only API with limited scope it's probably ok. But we should be able to reuse `SuspendGIL`, the machinery behind `allow_threads`, which is basically the same as `ThreadStateGuard` i think.